### ISSUE TITLE
Not all sub reads from multiple splits are written to FASTQ output file

### ISSUE THREAD

**Issue by callumparr** [nanoporetech/duplex-tools#22]
TL;DR is there some other filtering applied when writing sub reads to FASTQ, as some reads in middle_fasta do not end up in the final FASTQ. 

I have been using split_on_adapter as a standard to first split any reads that may result from failure to either split reads during sequencing or result from ligation. 

I allow for multiple splits and seems to work as expected.
```
duplex_tools split_on_adapter split_on_adapter_test/LambdaDNA_3/barcode11/input split_on_adapter_test/LambdaDNA_3/barcode11/NanoporeAdapter Native --debug_output --threads 12 --debug --print_alignment --allow_multiple_splits --trim_start 20 --trim_end 20 > split_on_adapter_test/LambdaDNA_3/barcode11/alignment.txt
```
so for instance 

2a0ed51b-3d0c-4877-b065-81f8d9be8739 becomes: 

```
@2a0ed51b-3d0c-4877-b065-81f8d9be8739_1 runid=e08a91d5edc5ac90e3d8c8b2faebd0af477edb35 sampleid=LamdaDNA_3 read=45762 ch=177 start_time=2021-11-13T20:10:00Z barcode=barcode11 0->3552
@2a0ed51b-3d0c-4877-b065-81f8d9be8739_2 runid=e08a91d5edc5ac90e3d8c8b2faebd0af477edb35 sampleid=LamdaDNA_3 read=45762 ch=177 start_time=2021-11-13T20:10:00Z barcode=barcode11 3590->3838
```

I have then adapted the adapters.py file so that it can search for my own internal adapters and then split the reads accordingly. These internal adapters are also created by ligation between two or more DNA fragments. Its like proximity ligation of fix chromatin with bridge adapters inserted in between ligation sites. 

```
duplex_tools split_on_adapter split_on_adapter_test/LambdaDNA_3/barcode11/NanoporeAdapter split_on_adapter_test/LambdaDNA_3/barcode11/APACseqAdapter Native --debug_output --threads 12 --debug --print_alignment --allow_multiple_splits --trim_start 20 --trim_end 20 > split_on_adapter_test/LambdaDNA_3/barcode11/alignment2.txt
```

Checking through some reads I can see where it makes a lot of sense it does a good job splitting reads into an expected number of subreads. Of course it may need some tuning but one issue is that although there might be several sequences created in the middle.fasta file for some reason some are not then written to the split.fastq.z file. 

Example might be here:

You can see a read that was first split by the default Nanopore sequencing adapters then one of the reads was further split as it found my custom adapter hence the _x_x readID nomenclature. 

```
012a9a4d-353c-4d94-a60c-ba20c4af3bdf_1
012a9a4d-353c-4d94-a60c-ba20c4af3bdf_2_1
012a9a4d-353c-4d94-a60c-ba20c4af3bdf_2_2
```

Or a read that did not contain an internal Nanopore sequencing adapter but did contain internal custom adapters and was split into 3 reads. 

```
0a9ac4fe-70c2-4c03-80a4-a0190dc4c153_1
0a9ac4fe-70c2-4c03-80a4-a0190dc4c153_2
0a9ac4fe-70c2-4c03-80a4-a0190dc4c153_3
```

And then there is some case where not all subreads are written out when it finds multiple custom adapters internally.

```
131b78a4-a711-43b2-b60a-1de3985c53bd_1
131b78a4-a711-43b2-b60a-1de3985c53bd_2
131b78a4-a711-43b2-b60a-1de3985c53bd_3
131b78a4-a711-43b2-b60a-1de3985c53bd_5
```

I noticed that it is missing because the sucecssive run broke between _3 and _5. 

Looking at the reads from the middle_fasta I am unsure why the fourth sequence is not written out. 

```
❯ grep "131b78a4-a711-43b2-b60a-1de3985c53bd" APACseqAdapter/fastq_runid_e08a91d5edc5ac90e3d8c8b2faebd0af477edb35_barcode11_pass_split_middle.fasta -A 1
>131b78a4-a711-43b2-b60a-1de3985c53bd
CTGTCTCTTATACACATCTCCTGCAGGAGATGTGTATAAGAGACAGACATGGGGATCATCCCGCCGAAGGAGTCCTTCCTGCCCCCCTGTTGCAGCAGGATCAGCCACGGACTTTGCCCGCCTGCAAGCTGCGTGGCCACGTCAAATTAATTGTGCAGGCAGCATACGCATGGCGGCTTTATACTGCCCGACGGAAATCCCCGCTTTCTGTGCGGCCAGCGCCTGTCGGCTCAGCGACTGTTCAACGACTGCCGCTGTTTTTTCGCATCACTTTCCGTACCAGAAAAATGACGCCTGACTCTGGCCATCTGCTCGTCAAATCTGGCCGCATCCAGACTCAAATCAACGACCAGATCGCCTACCGGTTCAGCCATACCGGACTCCTCCTGCGATCCTTCTGATACTGTCATCAGCATTACGTCATCCTCCGTCATGTCCGCCACATCGGGAAGCGGGATAACTTCATTCCGTCCGGGCAAAGCGGACACCTCCGGCAAGCCCTGCCGCTTTCTGCATCAG
>131b78a4-a711-43b2-b60a-1de3985c53bd
GTGTATAAGCTGGAACAGGCACACTTGAATCATGGCTTTATGACGTAACATCCGTTTGGGATGCGACTGCCACGGCCCGTGATTTCTCTGCCTTCGCGAGTTTTGAATGGTTCGCGGCGGCATTCATCCATCCATTCGGTAACGCGAATCGGATGATTACGGTCCTTGCGGTAAATCCGGCATGTACAGGATTCATTGTCCTGCTCAAAGTCCATGCCATCAAACTGCTGGTTTTCATTGATGATGCGGGACCAGCCATCAACGCCCACCACCGGAACGATGCCATTCTGCTTATCAGGAAAGGCGTAAATTTCTTTCGTCCACGGATTAAGGCCGTACTGGTTGGCAACGATCAGTAATGCGATGAACTGCACGTAGCTGGCATCACCTTTAAATGCCGTCTGGCGAAGTGGTGATCAGTTCCTGTGGGTCGACAGAATCCATGCCGACACGTTCAGCCAGCTTCCCAGCCAGCGTTGCGAGTGCAGTACTCATTCGTTTTATACCCTCTGAATCAATATCAACCTGGTGGTGAGCAATGGTTTCAACCATGTACCGGATGTGTTCTGCCATGCGCTCCTGAAACTCAACATCGTCATCAAACGCACGGGTAATGGATTTTTTGCTGGCCCCGTGGCGTTGCAAATGATCGATGCATAGCGATTCAAACAGGTGCTGGGGCAGGCCTTTTTCCATGTCGTCTGCCAGTTCTGCCT
>131b78a4-a711-43b2-b60a-1de3985c53bd
GCGCCCAGCCTTTCTGAGCCTCAAGACGATCCTGAATGTAATAGCGTTCATGGCTGAACTCCTGAAATAGCTGTGAAAATATCGCCCGCGAAATGCCGGGCTGATTAGGAAAACAGGAAAGGGGGTTAGTGAATGCTTTTGCTTGATCTCAGTTTCAGTATTAATATCCATTTTATAAGCGTCGACGGCTTCACGAAACATCTTTTCATCGCCAATAAAAGTGGCGATAGTGAATTTAGTCTGGATAGCCATAAGTGTTTGATCCATTCTTTGGGACTCCTGGCTGATTAAGTATGTCGATAAGGCGTTTCCATCCGTCGTAATTTACGGGTGATTCGTTCAAGTAAAGATTCGGAAGGGCAGCCAGCAACAGGCCACCCTGCAATGGCATATTGCATGGTGTGCTCCTTATTTATACATAACGAAAAACGCCTCGAGTGAAGCGTTATTGGTATGCGGTAAAACCGCACTCAGGCGGCCTTGATAGTCATATCATCTGAATCAAATATTCCTGATGTATCGATATCGGTAATTCTTATTCCTTCGCTACCATCCATTGGAGGCCATCCTTCCTGACCATTTCCATCATTCCAGTCGAACTCACACACAACACCATATGCATTAGAGTCGCTTGAAATTGCTATAAGCAGAGCATGTTGCGCCAGCATGATTAATACAGCATTTAATACAGAGCCGTGTTTATTGAGTCGGTATTCAGAGTCTGACCAGGTATTGATCTGGTGAAGTTTTTTCCTCTGTCATTACGTCATGGTCGATTTCAATTTCTATTGATGCTTTCCAGTCGTAATCAATGATGTATTTTTTGATGTTTGACATCTGTTCATATCCTCACAGATAAAAAATCGCCCTCACACTGGAGGGCAAAGAAGATTTCCATCAATCCAGAACAAGTCGGCTCCTGTTTAGTAACGAGCGACATTGCTCCGTGTATTCACTCGTTGGAATGAATACAC
>131b78a4-a711-43b2-b60a-1de3985c53bd
TTCCATCAATCCAGAACAAGTCGGCTCCTGTTTAGTAACGAGCGACATTGCTCCGTGTATTCACTCGTTGGAATGAATACACCTGTCTCTTATACACATCTCCTGCAGGAGATCTTATACACATCTCCTGCAGGAGATGTGTATAAGAGACGGATAACCTGCCGGACCACCGACTG
>131b78a4-a711-43b2-b60a-1de3985c53bd
CGAGCGACATTGCTCCGTGTATTCACTCGTTGGAATGAATACACCTGTCTCTTATACACATCTCCTGCAGGAGATGTGTATAAGAGACGGATAACCTGCCGGACCACCGACTG
```

I can see that there is some overlap between sub read 3 and 4 when written out to FASTQ. 

```
gunzip -c APACseqAdapter/fastq_runid_e08a91d5edc5ac90e3d8c8b2faebd0af477edb35_barcode11_pass_split_split.fastq.gz| grep "131b78a4-a711-43b2-b60a-1de3985c53bd" -A 1
@131b78a4-a711-43b2-b60a-1de3985c53bd_1 runid=e08a91d5edc5ac90e3d8c8b2faebd0af477edb35 sampleid=LamdaDNA_3 read=57364 ch=26 start_time=2021-11-13T21:51:20Z barcode=barcode11 0->519
CTGTCTCTTATACACATCTCCTGCAGGAGATGTGTATAAGAGACAGACATGGGGATCATCCCGCCGAAGGAGTCCTTCCTGCCCCCCTGTTGCAGCAGGATCAGCCACGGACTTTGCCCGCCTGCAAGCTGCGTGGCCACGTCAAATTAATTGTGCAGGCAGCATACGCATGGCGGCTTTATACTGCCCGACGGAAATCCCCGCTTTCTGTGCGGCCAGCGCCTGTCGGCTCAGCGACTGTTCAACGACTGCCGCTGTTTTTTCGCATCACTTTCCGTACCAGAAAAATGACGCCTGACTCTGGCCATCTGCTCGTCAAATCTGGCCGCATCCAGACTCAAATCAACGACCAGATCGCCTACCGGTTCAGCCATACCGGACTCCTCCTGCGATCCTTCTGATACTGTCATCAGCATTACGTCATCCTCCGTCATGTCCGCCACATCGGGAAGCGGGATAACTTCATTCCGTCCGGGCAAAGCGGACACCTCCGGCAAGCCCTGCCGCTTTCTGCATCAG
--
@131b78a4-a711-43b2-b60a-1de3985c53bd_2 runid=e08a91d5edc5ac90e3d8c8b2faebd0af477edb35 sampleid=LamdaDNA_3 read=57364 ch=26 start_time=2021-11-13T21:51:20Z barcode=barcode11 549->1265
GTGTATAAGCTGGAACAGGCACACTTGAATCATGGCTTTATGACGTAACATCCGTTTGGGATGCGACTGCCACGGCCCGTGATTTCTCTGCCTTCGCGAGTTTTGAATGGTTCGCGGCGGCATTCATCCATCCATTCGGTAACGCGAATCGGATGATTACGGTCCTTGCGGTAAATCCGGCATGTACAGGATTCATTGTCCTGCTCAAAGTCCATGCCATCAAACTGCTGGTTTTCATTGATGATGCGGGACCAGCCATCAACGCCCACCACCGGAACGATGCCATTCTGCTTATCAGGAAAGGCGTAAATTTCTTTCGTCCACGGATTAAGGCCGTACTGGTTGGCAACGATCAGTAATGCGATGAACTGCACGTAGCTGGCATCACCTTTAAATGCCGTCTGGCGAAGTGGTGATCAGTTCCTGTGGGTCGACAGAATCCATGCCGACACGTTCAGCCAGCTTCCCAGCCAGCGTTGCGAGTGCAGTACTCATTCGTTTTATACCCTCTGAATCAATATCAACCTGGTGGTGAGCAATGGTTTCAACCATGTACCGGATGTGTTCTGCCATGCGCTCCTGAAACTCAACATCGTCATCAAACGCACGGGTAATGGATTTTTTGCTGGCCCCGTGGCGTTGCAAATGATCGATGCATAGCGATTCAAACAGGTGCTGGGGCAGGCCTTTTTCCATGTCGTCTGCCAGTTCTGCCT
--
@131b78a4-a711-43b2-b60a-1de3985c53bd_3 runid=e08a91d5edc5ac90e3d8c8b2faebd0af477edb35 sampleid=LamdaDNA_3 read=57364 ch=26 start_time=2021-11-13T21:51:20Z barcode=barcode11 1298->2272
GCGCCCAGCCTTTCTGAGCCTCAAGACGATCCTGAATGTAATAGCGTTCATGGCTGAACTCCTGAAATAGCTGTGAAAATATCGCCCGCGAAATGCCGGGCTGATTAGGAAAACAGGAAAGGGGGTTAGTGAATGCTTTTGCTTGATCTCAGTTTCAGTATTAATATCCATTTTATAAGCGTCGACGGCTTCACGAAACATCTTTTCATCGCCAATAAAAGTGGCGATAGTGAATTTAGTCTGGATAGCCATAAGTGTTTGATCCATTCTTTGGGACTCCTGGCTGATTAAGTATGTCGATAAGGCGTTTCCATCCGTCGTAATTTACGGGTGATTCGTTCAAGTAAAGATTCGGAAGGGCAGCCAGCAACAGGCCACCCTGCAATGGCATATTGCATGGTGTGCTCCTTATTTATACATAACGAAAAACGCCTCGAGTGAAGCGTTATTGGTATGCGGTAAAACCGCACTCAGGCGGCCTTGATAGTCATATCATCTGAATCAAATATTCCTGATGTATCGATATCGGTAATTCTTATTCCTTCGCTACCATCCATTGGAGGCCATCCTTCCTGACCATTTCCATCATTCCAGTCGAACTCACACACAACACCATATGCATTAGAGTCGCTTGAAATTGCTATAAGCAGAGCATGTTGCGCCAGCATGATTAATACAGCATTTAATACAGAGCCGTGTTTATTGAGTCGGTATTCAGAGTCTGACCAGGTATTGATCTGGTGAAGTTTTTTCCTCTGTCATTACGTCATGGTCGATTTCAATTTCTATTGATGCTTTCCAGTCGTAATCAATGATGTATTTTTTGATGTTTGACATCTGTTCATATCCTCACAGATAAAAAATCGCCCTCACACTGGAGGGCAAAGAAGATTTCCATCAATCCAGAACAAGTCGGCTCCTGTTTAGTAACGAGCGACATTGCTCCGTGTATTCACTCGTTGGAATGAATACAC
--
@131b78a4-a711-43b2-b60a-1de3985c53bd_5 runid=e08a91d5edc5ac90e3d8c8b2faebd0af477edb35 sampleid=LamdaDNA_3 read=57364 ch=26 start_time=2021-11-13T21:51:20Z barcode=barcode11 2316->2341
GGATAACCTGCCGGACCACCGACTG

```

**Comment by onordesjo** [comment]
Hi @callumparr, 

The only reason a read would not be written out is if there's an overlapping match:

https://github.com/nanoporetech/duplex-tools/blob/master/duplex_tools/split_on_adapter.py#L173

We've decided to err on the side that doesn't cause any sequence duplication in the output in this case, but it's likely that we can solve it more elegantly somehow.

**Comment by callumparr** [comment]
Does overlapping match relate to the adapter sequences like below? 

<img width="528" alt="image" src="https://user-images.githubusercontent.com/41637575/192508792-cf249213-ccca-4a1b-8878-d584dc951aa3.png">

**Comment by onordesjo** [comment]
You can see which spans are candidates to write out by looking at `hits` just before this line:

https://github.com/nanoporetech/duplex-tools/blob/master/duplex_tools/split_on_adapter.py#L164


The idea is to write out the complement of the adapter hits, so the following loops through the intervals against the adapter, and the `hits` will be populated with the complement of this (the end of each hit + the beginning of the next hit).

Note that `0,0` and `len(seq),len(seq)` are added to allow for the first and last "complement" to the adapter hits to be written out.

```
for left_hit, right_hit in pairwise(
        [(0, 0), *result['locations'], (len(seq), len(seq))]):
    hits.append([left_hit[1], right_hit[0]])
```

I'd imagine seeing what that variable is would bring some clarity into why there's one read that's not being written out

**Comment by callumparr** [comment]
Thank you. For now, it seems by eye that there is some overlap.
